File: lookyswappy-app/database/models/Game.ts
import { Model, Query } from '@nozbe/watermelondb'
import { field, date, children } from '@nozbe/watermelondb/decorators'
import type GamePlayer from './GamePlayer'
import type Round from './Round'

export type GameStatus = 'active' | 'completed'

export default class Game extends Model {
  static table = 'games'
  static associations = {
    game_players: { type: 'has_many' as const, foreignKey: 'game_id' },
    rounds: { type: 'has_many' as const, foreignKey: 'game_id' },
  }

  @field('server_id') serverId!: string | null
  @field('name') name!: string | null
  @field('target_score') targetScore!: number
  @field('status') status!: GameStatus
  @field('winner_id') winnerId!: string | null
  @date('started_at') startedAt!: Date
  @date('ended_at') endedAt!: Date | null
  @date('last_modified') lastModified!: Date
  @field('is_synced') isSynced!: boolean

  @children('game_players') players!: Query<GamePlayer>
  @children('rounds') rounds!: Query<Round>
}

File: lookyswappy-api/app/models/game.py
import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import SyncableBase

if TYPE_CHECKING:
    from app.models.device import Device
    from app.models.player import GamePlayer
    from app.models.round import Round


class GameStatus(str, enum.Enum):
    ACTIVE = "active"
    COMPLETED = "completed"


class Game(SyncableBase):
    """A Lookyswappy game session."""

    __tablename__ = "games"

    device_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("devices.id"),
        index=True,
    )
    name: Mapped[str | None] = mapped_column(String(100), nullable=True)
    target_score: Mapped[int] = mapped_column(Integer, default=100)
    status: Mapped[GameStatus] = mapped_column(
        Enum(GameStatus),
        default=GameStatus.ACTIVE,
    )
    winner_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("game_players.id", use_alter=True),
        nullable=True,
    )
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
    )
    ended_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True),
        nullable=True,
    )

    # Relationships
    device: Mapped["Device"] = relationship(back_populates="games")
    players: Mapped[list["GamePlayer"]] = relationship(
        back_populates="game",
        cascade="all, delete-orphan",
        foreign_keys="GamePlayer.game_id",
    )
    rounds: Mapped[list["Round"]] = relationship(
        back_populates="game",
        cascade="all, delete-orphan",
    )
